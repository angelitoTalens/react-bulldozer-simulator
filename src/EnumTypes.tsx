export enum Direction {
    Up = "U",
    Down = "D",
    Left = "L",
    Right = "R",
}

export enum Commands {
    RotateLeft = "L",
    RotateRight = "R",
    Advance = "A"
}

export type Position = {
    X: number, 
    Y: number
}

export enum LandType {
    Plain = "o",
    Rocky = "r",
    TreePlanted = "t",
    Protected = "T",
    Cleared = ""
}

export enum SimulationStatus {
    Ongoing,
    NoGrid,
    OutOfBounds,
    ClearedProtected,
    UserTerminated
}

export enum TabKeys {
    Simulation = "Simulation",
    Results = "Results"
}

export type Cost = {
    quantity: number;
    value: number;
}

export type LandVisitSummary = {
    clearPlainLand: number;
    revisitClearedLand: number;
    clearRockyLand: number;
    clearTreePlantedLand: number;
}

export type Movements = {
    command: Commands;
    steps: number;
}