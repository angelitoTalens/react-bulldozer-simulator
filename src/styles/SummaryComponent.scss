.summary-container {
    padding-top: 20px;
}