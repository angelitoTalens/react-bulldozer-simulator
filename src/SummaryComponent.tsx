import React from "react";
import "./styles/SummaryComponent.scss"

import FuelUsageComponent from "./FuelUsageComponent";
import { LandVisitSummary, LandType, Cost, Movements, Commands } from "./EnumTypes";
import TotalCostComponent from "./TotalCostComponent";
import Container from "react-bootstrap/Container";
import Table from "react-bootstrap/Table";
import Row from "react-bootstrap/Row";

type SummaryComponentProps = {
    message: string;
    landTypeHistory: LandType[];
    movementHistory: Movements[];
    gridMap: LandType[][];
    paintDamageCount: number;
}

class SummaryComponent extends React.Component<SummaryComponentProps> {

    calculateTotalFuelUnits(summary: LandVisitSummary): number {
        let total = summary.clearPlainLand +
                    summary.revisitClearedLand +
                    summary.clearRockyLand * 2 +
                    summary.clearTreePlantedLand * 2;
        
        return total;
    }

    calculateLandVisits(landTypeHistory: LandType[]): LandVisitSummary {
        let summary: LandVisitSummary = {
            clearPlainLand: 0,
            revisitClearedLand: 0,
            clearRockyLand: 0,
            clearTreePlantedLand: 0,
        };

        for (let landType of landTypeHistory) {
            switch(landType) {
                case LandType.Plain:
                    summary.clearPlainLand++;
                    break;
                case LandType.Cleared:
                    summary.revisitClearedLand++;
                    break;
                case LandType.Rocky:
                    summary.clearRockyLand++;
                    break;
                case LandType.TreePlanted:
                    summary.clearTreePlantedLand++;
                    break;
                default: break;
            }
        }
        return summary;
    }

    calculateUnclearedBlock(grid: LandType[][]): number {
        let total = 0;
        for (let rowBlock of grid) {
            for (let block of rowBlock) {
                switch (block) {
                    case LandType.Plain:
                    case LandType.Rocky:
                    case LandType.TreePlanted:
                        total++;
                        break;
                    default: break;
                }
            }
        }
        return total;
    }

    renderFuelUsage(landTypeHistory: LandType[]) {
        const summary = this.calculateLandVisits(landTypeHistory);

        const clearPlainLandFuelUsage: Cost = {
            quantity: summary.clearPlainLand,
            value: summary.clearPlainLand
        }

        const clearRockyLandFuelUsage: Cost = {
            quantity: summary.clearRockyLand,
            value: summary.clearRockyLand * 2
        }

        const clearTreePlantedLandFuelUsage: Cost = {
            quantity: summary.clearTreePlantedLand,
            value: summary.clearTreePlantedLand * 2
        }

        const revisitClearedLandFuelUsage: Cost = {
            quantity: summary.revisitClearedLand,
            value: summary.revisitClearedLand
        }

        return(
            <FuelUsageComponent
                clearPlainLand={clearPlainLandFuelUsage}
                clearRockyLand={clearRockyLandFuelUsage}
                clearTreePlantedLand={clearTreePlantedLandFuelUsage}
                revisitClearedLand={revisitClearedLandFuelUsage}
            />
        );
    }

    renderTotalCost(
        landTypeHistory: LandType[],
        movementHistory: Movements[],
        gridMap: LandType[][],
        paintDamageCount: number
    ){
        const landVisits = this.calculateLandVisits(landTypeHistory);

        let clearProtectedTreeCount = 0;
        for (let landType of landTypeHistory) {
            if (landType === LandType.Protected) {
                clearProtectedTreeCount++;
            }
        }
      
        const totalFuel: Cost = {
            quantity: this.calculateTotalFuelUnits(landVisits),
            value: this.calculateTotalFuelUnits(landVisits)
        }

        const communication: Cost = {
            quantity: movementHistory.length,
            value: movementHistory.length
        }

        const unclearedBlock: Cost = {
            quantity: this.calculateUnclearedBlock(gridMap),
            value: this.calculateUnclearedBlock(gridMap) * 3
        }

        const clearProtectedTree: Cost = {
            quantity: clearProtectedTreeCount,
            value: clearProtectedTreeCount * 10
        }

        const paintDamage: Cost = {
            quantity: paintDamageCount,
            value: paintDamageCount * 2
        }

        return (
            <TotalCostComponent 
                totalFuel={totalFuel}
                communication={communication}
                unclearedBlock={unclearedBlock}
                clearProtectedTree={clearProtectedTree}
                paintDamage = {paintDamage}
            />
        );
    }

    renderMovementHistory(movementHistory: Movements[]) {
        let movements: string[] = [];
        for (let move of movementHistory) {
            (move.command === Commands.Advance) ? movements.push(move.command + move.steps + ", ") :
                                                  movements.push(move.command + ", ")
        }
        return (
            <Container>
                <Table size="sm" striped hover>
                    <thead>
                        <tr><th>Movements History</th></tr>
                    </thead>
                    <tbody>
                        <tr><td>{movements}</td></tr>
                    </tbody>
                </Table>
            </Container>
        );
    }

    renderSummary(message:string) {
        return (
            <Container className="summary-container">
                <Row> <h3>{message}</h3></Row>
                <Row>{this.renderMovementHistory(this.props.movementHistory)}</Row>
                <Row>{this.renderFuelUsage(this.props.landTypeHistory)}</Row>
                <Row>{this.renderTotalCost(
                        this.props.landTypeHistory,
                        this.props.movementHistory,
                        this.props.gridMap,
                        this.props.paintDamageCount)
                }</Row>
            </Container>
        );
    }

    render() {
        return (
            <div>
                {this.renderSummary(this.props.message)}
            </div>
        );
    }
}

export default SummaryComponent;