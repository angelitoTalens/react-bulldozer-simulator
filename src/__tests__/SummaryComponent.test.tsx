import React from 'react';
import renderer from 'react-test-renderer';

import SummaryComponent from '../SummaryComponent';
import { LandType, Movements, Commands } from '../EnumTypes';

test ("Test SummaryComponent", ()  => {

    const landTypeHistory: LandType[] = [
        LandType.Plain,
        LandType.Rocky,
        LandType.TreePlanted,
        LandType.Plain
    ];

    const movementHistory: Movements[] = [
        {command: Commands.Advance, steps: 1},
        {command: Commands.Advance, steps: 1},
        {command: Commands.RotateRight, steps: 0},
        {command: Commands.Advance, steps: 1},
        {command: Commands.RotateRight, steps: 0},
        {command: Commands.Advance, steps: 1},
    ];
    
    const gridMap: LandType[][] = [
        [LandType.Plain, LandType.Rocky],
        [LandType.TreePlanted, LandType.Plain],
    ];  

    const component = renderer.create(
        <SummaryComponent 
            message={"Test Summary"}
            landTypeHistory={landTypeHistory}
            movementHistory={movementHistory}
            gridMap={gridMap}
            paintDamageCount={0}
        />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

