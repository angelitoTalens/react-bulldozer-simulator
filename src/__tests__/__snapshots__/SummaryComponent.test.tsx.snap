// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SummaryComponent 1`] = `
<div>
  <div
    className="summary-container container"
  >
    <div
      className="row"
    >
       
      <h3>
        Test Summary
      </h3>
    </div>
    <div
      className="row"
    >
      <div
        className="container"
      >
        <table
          className="table table-sm table-striped table-hover"
        >
          <thead>
            <tr>
              <th>
                Movements History
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                A1, 
                A1, 
                R, 
                A1, 
                R, 
                A1, 
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="fuel-usage-container container"
      >
        <table
          className="table table-sm table-striped table-hover"
        >
          <thead>
            <tr>
              <th
                style={
                  Object {
                    "width": "40%",
                  }
                }
              >
                Activity
              </th>
              <th
                style={
                  Object {
                    "width": "30%",
                  }
                }
              >
                Visits
              </th>
              <th
                style={
                  Object {
                    "width": "30%",
                  }
                }
              >
                Fuel Usage (fuel units)
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                Clearing Plain Land
              </td>
              <td>
                2
              </td>
              <td>
                2
              </td>
            </tr>
            <tr>
              <td>
                Clearing Rocky Land
              </td>
              <td>
                1
              </td>
              <td>
                2
              </td>
            </tr>
            <tr>
              <td>
                Clearing Tree Planted Land
              </td>
              <td>
                1
              </td>
              <td>
                2
              </td>
            </tr>
            <tr>
              <td>
                Revisiting Cleared Land
              </td>
              <td>
                0
              </td>
              <td>
                0
              </td>
            </tr>
            <tr>
              <td>
                <b>
                  Total
                </b>
              </td>
              <td>
                <b>
                  4
                </b>
              </td>
              <td>
                <b>
                  6
                </b>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="total-cost-container container"
      >
        <table
          className="table table-sm table-striped table-hover"
        >
          <thead>
            <tr>
              <th
                style={
                  Object {
                    "width": "40%",
                  }
                }
              >
                Item
              </th>
              <th
                style={
                  Object {
                    "width": "30%",
                  }
                }
              >
                Quantity
              </th>
              <th
                style={
                  Object {
                    "width": "30%",
                  }
                }
              >
                Cost (Credit(s))
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                Communication Overhead
              </td>
              <td>
                6
              </td>
              <td>
                6
              </td>
            </tr>
            <tr>
              <td>
                Fuel Usage
              </td>
              <td>
                6
              </td>
              <td>
                6
              </td>
            </tr>
            <tr>
              <td>
                Uncleared Squares
              </td>
              <td>
                4
              </td>
              <td>
                12
              </td>
            </tr>
            <tr>
              <td>
                Cleared Preserved Tree(s)
              </td>
              <td>
                0
              </td>
              <td>
                0
              </td>
            </tr>
            <tr>
              <td>
                Paint Damage Cost
              </td>
              <td>
                0
              </td>
              <td>
                0
              </td>
            </tr>
            <tr>
              <td>
                <b>
                  Total Cost
                </b>
              </td>
              <td />
              <td>
                <b>
                  24
                </b>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
